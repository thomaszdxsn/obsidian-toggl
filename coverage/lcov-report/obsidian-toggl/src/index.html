
<!doctype html>
<html lang="en">

<head>
    <title>Code coverage report for obsidian-toggl/src</title>
    <meta charset="utf-8" />
    <link rel="stylesheet" href="../../prettify.css" />
    <link rel="stylesheet" href="../../base.css" />
    <link rel="shortcut icon" type="image/x-icon" href="../../favicon.png" />
    <meta name="viewport" content="width=device-width, initial-scale=1" />
    <style type='text/css'>
        .coverage-summary .sorter {
            background-image: url(../../sort-arrow-sprite.png);
        }
    </style>
</head>
    
<body>
<div class='wrapper'>
    <div class='pad1'>
        <h1><a href="../../index.html">All files</a> obsidian-toggl/src</h1>
        <div class='clearfix'>
            
            <div class='fl pad1y space-right2'>
                <span class="strong">56.74% </span>
                <span class="quiet">Statements</span>
                <span class='fraction'>101/178</span>
            </div>
        
            
            <div class='fl pad1y space-right2'>
                <span class="strong">30.76% </span>
                <span class="quiet">Branches</span>
                <span class='fraction'>20/65</span>
            </div>
        
            
            <div class='fl pad1y space-right2'>
                <span class="strong">28.3% </span>
                <span class="quiet">Functions</span>
                <span class='fraction'>15/53</span>
            </div>
        
            
            <div class='fl pad1y space-right2'>
                <span class="strong">55.62% </span>
                <span class="quiet">Lines</span>
                <span class='fraction'>84/151</span>
            </div>
        
            
        </div>
        <p class="quiet">
            Press <em>n</em> or <em>j</em> to go to the next uncovered block, <em>b</em>, <em>p</em> or <em>k</em> for the previous block.
        </p>
        <template id="filterTemplate">
            <div class="quiet">
                Filter:
                <input oninput="onInput()" type="search" id="fileSearch">
            </div>
        </template>
    </div>
    <div class='status-line medium'></div>
    <div class="pad1">
<table class="coverage-summary">
<thead>
<tr>
   <th data-col="file" data-fmt="html" data-html="true" class="file">File</th>
   <th data-col="pic" data-type="number" data-fmt="html" data-html="true" class="pic"></th>
   <th data-col="statements" data-type="number" data-fmt="pct" class="pct">Statements</th>
   <th data-col="statements_raw" data-type="number" data-fmt="html" class="abs"></th>
   <th data-col="branches" data-type="number" data-fmt="pct" class="pct">Branches</th>
   <th data-col="branches_raw" data-type="number" data-fmt="html" class="abs"></th>
   <th data-col="functions" data-type="number" data-fmt="pct" class="pct">Functions</th>
   <th data-col="functions_raw" data-type="number" data-fmt="html" class="abs"></th>
   <th data-col="lines" data-type="number" data-fmt="pct" class="pct">Lines</th>
   <th data-col="lines_raw" data-type="number" data-fmt="html" class="abs"></th>
</tr>
</thead>
<tbody><tr>
	<td class="file medium" data-value="atoms.ts"><a href="atoms.ts.html">atoms.ts</a></td>
	<td data-value="63.33" class="pic medium">
	<div class="chart"><div class="cover-fill" style="width: 63%"></div><div class="cover-empty" style="width: 37%"></div></div>
	</td>
	<td data-value="63.33" class="pct medium">63.33%</td>
	<td data-value="60" class="abs medium">38/60</td>
	<td data-value="43.75" class="pct low">43.75%</td>
	<td data-value="32" class="abs low">14/32</td>
	<td data-value="31.25" class="pct low">31.25%</td>
	<td data-value="16" class="abs low">5/16</td>
	<td data-value="65.45" class="pct medium">65.45%</td>
	<td data-value="55" class="abs medium">36/55</td>
	</tr>

<tr>
	<td class="file low" data-value="hooks.ts"><a href="hooks.ts.html">hooks.ts</a></td>
	<td data-value="30.26" class="pic low">
	<div class="chart"><div class="cover-fill" style="width: 30%"></div><div class="cover-empty" style="width: 70%"></div></div>
	</td>
	<td data-value="30.26" class="pct low">30.26%</td>
	<td data-value="76" class="abs low">23/76</td>
	<td data-value="0" class="pct low">0%</td>
	<td data-value="26" class="abs low">0/26</td>
	<td data-value="7.4" class="pct low">7.4%</td>
	<td data-value="27" class="abs low">2/27</td>
	<td data-value="26.98" class="pct low">26.98%</td>
	<td data-value="63" class="abs low">17/63</td>
	</tr>

<tr>
	<td class="file high" data-value="utils.tsx"><a href="utils.tsx.html">utils.tsx</a></td>
	<td data-value="95.23" class="pic high">
	<div class="chart"><div class="cover-fill" style="width: 95%"></div><div class="cover-empty" style="width: 5%"></div></div>
	</td>
	<td data-value="95.23" class="pct high">95.23%</td>
	<td data-value="42" class="abs high">40/42</td>
	<td data-value="85.71" class="pct high">85.71%</td>
	<td data-value="7" class="abs high">6/7</td>
	<td data-value="80" class="pct high">80%</td>
	<td data-value="10" class="abs high">8/10</td>
	<td data-value="93.93" class="pct high">93.93%</td>
	<td data-value="33" class="abs high">31/33</td>
	</tr>

</tbody>
</table>
</div>
                <div class='push'></div><!-- for sticky footer -->
            </div><!-- /wrapper -->
            <div class='footer quiet pad2 space-top1 center small'>
                Code coverage generated by
                <a href="https://istanbul.js.org/" target="_blank" rel="noopener noreferrer">istanbul</a>
                at 2024-03-30T14:07:17.202Z
            </div>
        <script src="../../prettify.js"></script>
        <script>
            window.onload = function () {
                prettyPrint();
            };
        </script>
        <script src="../../sorter.js"></script>
        <script src="../../block-navigation.js"></script>
    </body>
</html>
    