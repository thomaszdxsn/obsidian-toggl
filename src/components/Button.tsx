import { css } from '@emotion/css'
import clsx from 'clsx'
import React from 'react'
import { ClipLoader } from 'react-spinners'

interface Props extends React.ButtonHTMLAttributes<HTMLButtonElement> {
  children: React.ReactNode
  isLoading?: boolean
  loadingIconSize?: number
}


export const Button = ({ children, isLoading, loadingIconSize = 13, className, ...props }: Props) => {
  return <button
    {...props}
    disabled={isLoading || props.disabled}
    className={
      clsx(css`
        cursor: ${isLoading ? "wait" : "pointer"};
      `, className)
    }
  >{isLoading ?
    <ClipLoader size={loadingIconSize}
      data-testid="loading-icon"
      cssOverride={{
        borderColor: "var(--accent-h)",
      }}
    />
    : children}</button>
}