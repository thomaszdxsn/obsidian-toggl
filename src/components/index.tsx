export { CurrentTimer } from './CurrentTimer'
export { MainView } from './MainView'
export { Provider } from './Provider'
export { TimerCard } from './TimerCard'
export { TimerList } from './TimerList'
export { TimerForm } from './TimerForm'
export { Button } from './Button'
export { Timeline } from './Timeline'