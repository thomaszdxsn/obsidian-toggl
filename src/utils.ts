import { TimeEntry, Timer } from "./interfaces"

export const formatSeconds = (seconds: number) => {
	const hours = Math.floor(seconds / 3600)
	const minutes = Math.floor((seconds - hours * 3600) / 60)
	const second = seconds - hours * 3600 - minutes * 60
	return `${hours.toString().padStart(2, "0")}:${minutes.toString().padStart(2, "0")}:${second.toString().padStart(2, "0")}`
}


export const nowPercentageInDay = () => {
	const now = new Date()
	const secondsInDay = 24 * 60 * 60
	const seconds = now.getHours() * 60 * 60 + now.getMinutes() * 60 + now.getSeconds()
	return Number((seconds / secondsInDay).toFixed(2))
}


export const isSameTimer = ({ timer, entry }: { timer: Timer, entry: TimeEntry }) => {
	return timer.projectId === entry.project_id && timer.description === entry.description && timer.tagIds.join(",") === entry.tag_ids.join(",")
}