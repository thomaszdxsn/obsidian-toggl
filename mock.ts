import { faker } from '@faker-js/faker'
import { Project, TimeEntry } from 'src/interfaces'


export function createRandomTimeEntry(): TimeEntry {
  return {
    id: faker.number.int(),
    pid: faker.number.int(),
    project_id: faker.number.int(),
    description: faker.lorem.sentence(),
    duronly: faker.datatype.boolean(),
    wid: faker.number.int(),
    tags: [faker.lorem.word()],
    tag_ids: [faker.number.int()],
    start: faker.date.recent().toISOString(),
    stop: faker.date.recent().toISOString(),
    duration: faker.number.int(),
    billable: faker.datatype.boolean(),
    at: faker.date.recent().toISOString(),
    server_deleted_at: faker.date.recent().toISOString(),
    user_id: faker.number.int(),
    task_id: faker.number.int(),
    uid: faker.number.int(),
    tid: faker.number.int(),
    workspace_id: faker.number.int(),
  }
}

export function createRandomProject(): Project {
  return {
    id: faker.number.int(),
    wid: faker.number.int(),
    cid: faker.number.int(),
    name: faker.lorem.word(),
    billable: faker.datatype.boolean(),
    is_private: faker.datatype.boolean(),
    active: faker.datatype.boolean(),
    at: faker.date.recent().toISOString(),
    created_at: faker.date.recent().toISOString(),
    color: faker.internet.color(),
    auto_estimates: faker.datatype.boolean(),
    actual_hours: faker.number.int(),
    rate: faker.number.float(),
    template: faker.datatype.boolean(),
    template_id: faker.number.int(),
    actual_seconds: faker.number.int(),
    client_id: faker.number.int(),
    currency: faker.finance.currencyCode(),
    current_period: {
      at: faker.date.recent().toISOString(),
    },
    end_date: faker.date.recent().toISOString(),
    estimated_hours: faker.number.int(),
    estimated_seconds: faker.number.int(),
    fixed_fee: faker.number.float(),
    rate_last_updated: faker.date.recent().toISOString(),
    recurring: faker.datatype.boolean(),
    start_date: faker.date.recent().toISOString(),
    server_deleted_at: faker.date.recent().toISOString(),
    status: faker.lorem.word(),
    workspace_id: faker.number.int(),
  }
}
