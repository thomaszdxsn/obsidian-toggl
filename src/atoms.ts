import { atom, createStore } from "jotai";
import type { Me, TimeEntry, Timer } from "./interfaces";
import dayjs from "dayjs";
import { formatSeconds } from "./utils";


export const store = createStore();

export const savedTimersAtom = atom<Timer[]>([])

export const currentEntryAtom = atom<TimeEntry | null>(null)

export const passedSecondsAtom = atom<number | null>(null)

export const passedTimeAtom = atom<string | null>(get => {
	const passedSeconds = get(passedSecondsAtom)
	if (!passedSeconds) {
		return null
	}
	return formatSeconds(passedSeconds)
})

export const meAtom = atom<Me | null>(null)

export const tagsAtom = atom<Me["tags"]>(get => {
	const me = get(meAtom)
	return me?.tags ?? []
})

export const projectsAtom = atom<Me["projects"]>(get => {
	const me = get(meAtom)
	return me?.projects ?? []
})

export const activeProjectsAtom = atom(get => {
	const projects = get(projectsAtom)
	return projects.filter(project => project.active)
})

export const projectDictAtom = atom(get => {
	const projects = get(projectsAtom)
	return Object.fromEntries(projects.map(project => [project.id, project]))
})

export const currentEntryProjectAtom = atom(get => {
	const currentEntry = get(currentEntryAtom)
	const projects = get(projectsAtom)
	if (!currentEntry) {
		return null
	}
	return projects.find(project => project.id === currentEntry.project_id)
})

export const tick = () => {
	const interval = setInterval(() => {
		const currentEntry = store.get(currentEntryAtom)
		if (!currentEntry) {
			store.set(passedSecondsAtom, null)
			return
		}
		const passedSeconds = dayjs().diff(dayjs(currentEntry.start), 'second')
		store.set(passedSecondsAtom, passedSeconds)
	}, 1000)
	return interval
}

const myTimeEntriesAtom = atom(get => {
	const me = get(meAtom)
	return me?.time_entries ?? []
})

export const todayTimeEntriesAtom = atom(get => {
	const entries = get(myTimeEntriesAtom)
	return entries.filter(entry => dayjs().isSame(entry.start, 'day'))
})
