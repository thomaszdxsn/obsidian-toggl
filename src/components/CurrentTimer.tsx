import React from "react";
import { useAtom, useAtomValue, useSetAtom } from "jotai";
import { useStopTimerMutation } from "../hooks";
import {
  currentEntryAtom,
  currentEntryProjectAtom,
  currentTimerAtom,
  passedSecondsAtom,
  passedTimeAtom,
} from "../atoms";
import { css } from "@emotion/css";
import { FiPause } from "react-icons/fi";
import { TimeDisplay } from "./TimeDisplay";
import { Button } from "./Button";
import { generateTimerId } from "../utils";

export const CurrentTimer = ({
  placeholder = "No time entry running",
}: {
  placeholder?: string;
}) => {
  const passed = useAtomValue(passedTimeAtom);
  const setPassedNumbers = useSetAtom(passedSecondsAtom);
  const [currentEntry, setCurrentEntry] = useAtom(currentEntryAtom);
  const currentTimer = useAtomValue(currentTimerAtom);
  const stopTimerMutation = useStopTimerMutation(
    currentTimer ? generateTimerId(currentTimer) : undefined,
  );
  const currentProject = useAtomValue(currentEntryProjectAtom);
  const onStop = () => {
    if (!currentEntry) {
      return;
    }
    stopTimerMutation.mutate(
      { timeEntryId: currentEntry.id, workspaceId: currentEntry.workspace_id },
      {
        onSuccess: () => {
          setCurrentEntry(null);
          setPassedNumbers(null);
        },
      },
    );
  };
  if (!passed) {
    return (
      <div
        className={css`
          height: 100px;
          display: flex;
          justify-content: center;
          align-items: center;
          border-radius: var(--radius-m);
          background-color: var(--background-primary);
        `}
      >
        {placeholder}
      </div>
    );
  }
  return (
    <div
      className={css`
        display: flex;
        flex-direction: column;
        background-color: var(--background-primary);
        padding: var(--size-4-4);
        border-radius: var(--radius-m);
        height: 100px;
        container-type: inline-size;
      `}
    >
      <Button
        onClick={onStop}
        isLoading={stopTimerMutation.isPending}
        className={css`
          flex: 1;
          cursor: ${stopTimerMutation.isPending ? "wait" : "pointer"};
          margin-bottom: var(--size-4-2);
        `}
      >
        <FiPause />
      </Button>
      <div
        className={css`
          display: flex;
          justify-content: center;
          align-items: center;
          flex-wrap: wrap;
          @container (min-width: 200px) {
            justify-content: space-between;
          }
        `}
      >
        <div>
          {currentProject && (
            <div style={{ color: currentProject.color }}>
              {currentProject.name}
            </div>
          )}
          {currentEntry && (
            <div
              className={css`
                font-size: var(--font-smallest);
                color: var(--text-muted);
              `}
            >
              {currentEntry.description}
            </div>
          )}
        </div>
        <TimeDisplay
          className={css`
            font-size: var(--font-ui-large);
            font-weight: var(--font-semibold);
          `}
        >
          {passed}
        </TimeDisplay>
      </div>
    </div>
  );
};
