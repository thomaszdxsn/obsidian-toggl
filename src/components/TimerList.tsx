import React from 'react'
import { css } from "@emotion/css"
import { useAtomValue } from "jotai"
import { useMemo } from "react"
import { projectsAtom, savedTimersAtom } from "../atoms"
import { Project } from "../interfaces"
import TogglPlugin from "../main"
import { TimerCard } from './TimerCard'
import clsx from 'clsx'

interface Props {
  onSave?: () => void
  plugin: TogglPlugin
}


export const TimerList = ({ onSave, plugin }: Props) => {
  const timers = useAtomValue(savedTimersAtom)
  const projects = useAtomValue(projectsAtom)
  const projectDict = useMemo(() => Object.fromEntries(
    projects.map(project => [project.id, project])
  ), [projects])
  const findProject = (projectId: number): Project | undefined => projectDict[projectId]

  return (
    <div className={css`
      container-type: inline-size;
      container-name: timer-list;
    `}>
    <div className={clsx("timer-list", css`
      display: grid;
      gap: var(--size-4-1);

      @container timer-list (min-width: 250px) {
        grid-template-columns: repeat(2, minmax(0, 1fr));
      }
    `)}>
      {timers.map((timer, index) => {
        const project = findProject(timer.projectId)
        return <TimerCard onSuccess={onSave} key={index} timer={timer} project={project} plugin={plugin} />
      })}
    </div>
    </div>
  )
}