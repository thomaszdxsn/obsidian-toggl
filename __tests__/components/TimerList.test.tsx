import { TimerList } from "../../src/components/TimerList";
import "@testing-library/jest-dom";
import React from 'react';
import { render } from '@testing-library/react';
import { createRandomTimer, createRandomMe, createRandomProject } from "../../mock";
import { store, meAtom, savedTimersAtom } from "../../src/atoms";
import { TestingProvider } from "../../src/utils";


it("render correctly", () => {
  const project = createRandomProject({
    name: "Project Name",
    color: "red",
  })
  const timers = [createRandomTimer({
    projectId: project.id,
    projectName: project.name,
    description: "123",
    tags: []
  })]
  const me = createRandomMe({
    projects: [project]
  })
  store.set(meAtom, me)
  store.set(savedTimersAtom, timers)
  const {container, getByText} = render(
    <TestingProvider>
    <TimerList plugin={{} as any} />
    </TestingProvider>
  )
  const card = getByText("Project Name")
  expect(card).toBeInTheDocument()
  expect(container).toMatchSnapshot()
})