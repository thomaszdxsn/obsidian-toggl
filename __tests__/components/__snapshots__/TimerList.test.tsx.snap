// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render correctly 1`] = `
<div>
  <div
    class="css-1jqry03"
  >
    <section
      class="css-18foawe"
    >
      <div
        class="css-1pqyin1"
      >
        <div
          style="color: rgb(71, 119, 2);"
        >
          Project Name
        </div>
        <div
          class="css-y89s61"
        >
          Candidus caste cerno combibo pecto coruscus adiuvo timor crebro.
        </div>
        <div
          class="css-1pbu1sv"
        >
          reprehenderit
        </div>
      </div>
      <div
        class="css-1onrrkk"
      >
        <div
          class="css-4dyp2g css-ltzhbi"
        />
        <button
          class="css-1440elo"
        >
          <svg
            fill="none"
            height="1em"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            viewBox="0 0 24 24"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <polygon
              points="5 3 19 12 5 21 5 3"
            />
          </svg>
        </button>
      </div>
    </section>
  </div>
</div>
`;
